strawberryfields/io/xir_io.py

```diff
@@ -0,0 +1,294 @@
+# Copyright 2019-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains functions for loading and saving Strawberry Fields
+:class:`~.Program` objects from/to Blackbird scripts and Strawberry Fields
+code.
+"""
+# pylint: disable=protected-access,too-many-nested-blocks
+from decimal import Decimal
+from typing import Iterable
+
+import numpy as np
+
+import xir
+from xir.program import Declaration
+
+import strawberryfields.program as sfp
+from strawberryfields.tdm.tdmprogram import TDMProgram
+import strawberryfields.parameters as sfpar
+from strawberryfields import ops
+
+
+# pylint: disable=too-many-branches
+def from_xir(xir_prog):
+    """Convert an XIR Program to a Strawberry Fields program.
+
+    Args:
+        xir_prog (xir.Program): the input XIR program object
+
+    Returns:
+        Program: corresponding Strawberry Fields program
+    """
+    # only script-level statements are part of `xir_prog.statements`, which can only have integer
+    # wires, leading to `xir_prog.wires` only containing integer wire labels
+    num_of_modes = int(max(xir_prog.wires or [-1])) + 1
+    name = xir_prog.options.get("name", "xir")
+    if num_of_modes == 0:
+        raise ValueError(
+            "The XIR program is empty and cannot be transformed into a Strawberry Fields program"
+        )
+    prog = sfp.Program(num_of_modes, name=name)
+
+    # append the quantum operations
+    with prog.context as q:
+        for op in xir_prog.statements:
+            # check if operation name is in the list of
+            # defined StrawberryFields operations.
+            # This is used by checking against the ops.py __all__
+            # module attribute, which contains the names
+            # of all defined quantum operations
```

Comment:
Does this mean the converter will fail if a user-defined gate is applied in the XIR program?

Reply:
Yes, I've fixed this now so that all gate definitions are automatically expanded before conversion.